### PR TITLE
Fixes #12: Add new pipeline wizard with create and edit functionality

### LINKED ISSUE #12: Build onboarding workflow/wizard for defining data interface + pipeline
Multi-step form to take in from the user:

## First screen

- [x] Pipeline Name (text input)
- [x] Data Input Type (dropdown)
- [x] Database Connection String (large text field)

## Second screen
- [x] Script name (text input)
- [x] Script (monaco field)

## (on save)

- [x] Auto-generate the step order (1 or next available step order for that pipeline) -- can only create one for one, so `1` is fine
- [x] Save all the model objects to the database
- [x] Redirect user to the pipeline list page showing "Your pipeline was successfully saved" message

## Other considerations

- [x] Supports forward/back buttons, fill with the previous provided information
- [x] Don't commit to the database until the end

- Sample code: https://stackoverflow.com/questions/14901680/how-to-do-a-multi-step-form-in-django
- Alternative option if DIY becomes complex: https://www.imagescape.com/blog/multipage-forms-django/

## Related tasks
- #13
- #14

### PR DESCRIPTION
**Pull request by wfzyx**
Creates "Pipeline list" screen

How to test:

docker compose up
http://0.0.0.0:8000/projects/project-id-here/new-pipeline/

Closes #12

### CODE REVIEW

File: etlman/projects/authorizers.py [hunk 1]
```diff
@@ -6,7 +6,7 @@ def user_is_authenticated(request: HttpRequest, **view_kwargs: dict) -> bool:
 
 
 def user_is_project_collaborator(request: HttpRequest, **view_kwargs: dict) -> bool:
-    project_id = view_kwargs["project_id"]
+    project_id = str(view_kwargs["project_id"])
```
Comment (tobiasmcnulty) [inline etlman/projects/authorizers.py]: Why is this needed?

File: etlman/projects/views.py [hunk 2]
```diff
@@ -76,3 +75,135 @@ def new_project_view(request):
...
+    project = get_object_or_404(Project, pk=project_id)
+    username = request.user.username
+    # Form functionality
+    if request.method == "POST":
+        form_pipeline = PipelineForm(request.POST)
+        form_datainterface = DataInterfaceForm(request.POST)
+        if form_pipeline.is_valid() and form_datainterface.is_valid():
+
+            filled_datainterface = form_datainterface.save(commit=False)
+            filled_datainterface.name = request.POST.getlist("name")[-1]
```
Comment (tobiasmcnulty) [inline etlman/projects/views.py]: Please use the form `prefix`

File: etlman/projects/views.py [hunk 3]
```diff
@@ -76,3 +75,135 @@ def new_project_view(request):
...
+            initial={
+                "name": form_pipeline["name"] + "_script",
+                "script": "def main():\n\t...\nif __name__ == '__main__':\n\tmain()",
+            },
+        )
+    context = {"form_step": form_step, "project": project}
+    return render(request, "projects/new_step.html", context)
+
+
+def clear_session(request):
```
Comment (tobiasmcnulty) [inline etlman/projects/views.py]: `clear_step_wizard_session_variables()` perhaps?

File: etlman/projects/views.py [hunk 4]
```diff
@@ -76,3 +75,135 @@ def new_project_view(request):
...
+            filled_pipeline = form_pipeline.save(commit=False)
+            filled_pipeline.name = request.POST.getlist("name")[0]
+            filled_pipeline.input_id = filled_datainterface.pk
+            request.session["pipeline"] = model_to_dict(filled_pipeline)
+
+            return HttpResponseRedirect(
+                reverse("projects:new_step", args=(project.pk,))
+            )
+    else:  # GET
+        # If "in_transaction" is on session, the user came back from multiform step2
```
Comment (tobiasmcnulty) [inline etlman/projects/views.py]: Right now if you refresh page one after going back from page 2, it clears the values of the form. It would be nice to persist that here. The more I think about it, it might not even be necessary to clear the values in case someone comes back later. Some web apps (github for example) will save the contents of this text field even if I close it and come back.

File: etlman/projects/views.py [hunk 5]
```diff
@@ -76,3 +75,135 @@ def new_project_view(request):
...
+    if request.method == "POST":
+        form_step = NewStepForm(request.POST, instance=step)
+        if "save" in request.POST and form_step.is_valid():
+
+            # New Data Interface
+            saved_data_interface = DataInterface.objects.create(
+                # # project=project,
+                # {**form_data_interface},
+                project=project,  # Project(*form_data_interface)
+                name=form_data_interface["name"],
```
Comment (tobiasmcnulty) [inline etlman/projects/views.py]: Shorten this as discussed

File: etlman/projects/authorizers.py (hunk 1 above)
Comment (wfzyx) [reply etlman/projects/authorizers.py]: This is due to a mypy complaint

`etlman/projects/authorizers.py:12: error: Incompatible type for lookup 'pk': (got "Dict[Any, Any]", expected "Union[str, int]")`

So it is expecting a int or str, but is getting a dict[any, any]

File: etlman/projects/views.py (hunk 2 above)
Comment (wfzyx) [reply etlman/projects/views.py]: This was solved after the review, please check updated version

File: etlman/projects/views.py (hunk 3 above)
Comment (wfzyx) [reply etlman/projects/views.py]: Done!

File: etlman/projects/views.py (hunk 4 above)
Comment (wfzyx) [reply etlman/projects/views.py]: Done!

File: etlman/projects/views.py (hunk 5 above)
Comment (wfzyx) [reply etlman/projects/views.py]: We tried multiple variations of collection destruction here, but because the end result has more fields than what the form is expecting it, it errors out.

To be specific, it has "id" as extra.

So is it still worth it to go `{k:v for k, v in **form_data_interface if k is not 'id"}` or something similar ?

File: etlman/templates/projects/new_step.html [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        {% csrf_token %}
+        {{ form_step|crispy }}
+        <div class="row pt-3">
+            <div class="col">
+                <button type="submit" name="back" class="btn btn-danger">Back</button>
+            </div>
+            <div class="col">
+                <div class="control-group">
+                    <div class="d-flex flex-row-reverse">
+                        <button type="submit" name="save" class="btn btn-primary">Create Pipeline</button>
```
Comment (tobiasmcnulty) [inline etlman/templates/projects/new_step.html]: Since the form can create or edit:
```suggestion
                        <button type="submit" name="save" class="btn btn-primary">Save Pipeline</button>
```

File: etlman/templates/projects/new_pipeline.html [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+            </div>
+            <div class="col">
+                <div class="control-group">
+                    <div class="d-flex flex-row-reverse">
+                        <button type="submit" class="btn btn-primary">Next</button>
+                    </div>
+                </div>
+            </div>
+        </div>
+</div>
```
Comment (tobiasmcnulty) [inline etlman/templates/projects/new_pipeline.html]:
```suggestion
```

File: etlman/projects/forms.py [hunk 8]
```diff
@@ -1,24 +1,63 @@
 from django import forms
 
-from .models import Project, Step
+from .models import DataInterface, Pipeline, Project, Step
+
+# from django.utils.http import urlencode
```
Comment (tobiasmcnulty) [inline etlman/projects/forms.py]: The commented out code here and below can be removed,

File: etlman/projects/admin.py [hunk 9]
```diff
@@ -17,7 +17,7 @@ class StepInline(admin.TabularInline):
 
 class ProjectAdmin(admin.ModelAdmin):
     inlines = [CollaboratorInline, DataInterfaceInLine]
-    list_display = ["name", "description"]
+    list_display = ["id", "name", "description"]
```
Comment (tobiasmcnulty) [inline etlman/projects/admin.py]: Nitpick (ID becomes the link and is very small to click on):
```suggestion
    list_display = ["name", "description"]
```

(If there is more to this than I'm getting please push back!)

File: etlman/projects/admin.py [hunk 10]
```diff
@@ -26,7 +26,7 @@ class DataInterfaceAdmin(admin.ModelAdmin):
 
 
 class PipelineAdmin(admin.ModelAdmin):
-    list_display = ["name", "project"]
+    list_display = ["id", "name", "project"]
```
Comment (tobiasmcnulty) [inline etlman/projects/admin.py]:
```suggestion
    list_display = ["name", "project"]
```

File: etlman/projects/tests/test_views/test_view_pipeline.py [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+        assert Step.objects.count() == 1
+        assert SessionKeyEnum.STEP.value in nonadmin_client.session
+        assert (
+            nonadmin_client.session[SessionKeyEnum.STEP.value]["name"] == data["name"]
+        )
+        assert pipeline.name in str(response.content)
+
+        # TODO:
+        # - Verify status code (not sure what it should be)
+        # - Verify pipeline name in form or HTML
```
Comment (tobiasmcnulty) [inline etlman/projects/tests/test_views/test_view_pipeline.py]:
```suggestion
```

**Review by tobiasmcnulty — APPROVED** [review]
🧙🏻

File: etlman/projects/admin.py (hunk 9 above)
Comment (ronardcaktus) [reply etlman/projects/admin.py]: I included them because prior to implementing link buttons, I wanted to see the id of the object. I am fine with the removal.
